### PR TITLE
feat(training): add optional EMA weight averaging

### PR DESCRIPTION
**Pull request by KakaruHayate**
## Summary

Ports the EMA lifecycle from `openvpi/DiffSinger:v3-train` into the current training framework without pulling in the unfinished v3 rewrite, then hardens it for strict checkpoint handling and teacher-student training.

- update EMA after every real optimizer step
- validate/test with EMA weights, then restore optimizer-owned weights in-place
- persist EMA shadows and configuration metadata separately
- support include/exclude glob filters
- reject ambiguous resume/finetune operations by default
- provide a stateless, stop-gradient EMA teacher forward for consistency/Self-Flow-style objectives

## Strict loading policy

- EMA enabled + checkpoint without EMA: rejected by default; requires explicit `ema_missing_state: initialize`
- legacy EMA checkpoint without metadata: rejected by default; requires explicit `ema_missing_config: allow`
- EMA keys, tensor shapes, decay, include patterns, and exclude patterns are checked on strict resume
- finetuning explicitly selects `finetune_weight_source: model|ema`; requesting unavailable EMA weights is rejected
- EMA finetuning overlays tracked EMA parameters on the complete regular model state

## Teacher forward

`BaseTask.ema_teacher_forward(...)` runs `self.model` with EMA shadow parameters through `torch.func.functional_call`. It uses eval mode and no gradients, restores the student training mode, and never replaces student parameters or optimizer references. The task remains responsible for teacher inputs, intermediate feature extraction, and the consistency/Self-Flow loss.

## Configuration

```yaml
finetune_weight_source: "model"  # "model" or "ema"
ema_enabled: false
ema_decay: 0.999
ema_include_params:/n  - "model.*"
ema_exclude_params: []
ema_missing_state: "error"       # "error" or "initialize"
ema_missing_config: "error"      # "error" or "allow"
```

## Validation

Focused tests were run externally and removed afterward to follow the repository convention of not retaining test files.

- original 7 EMA lifecycle/serialization checks passed
- 10 strict-loading, finetune-source, namespace, eval-mode, and stateless teacher checks passed
- `python -m compileall`: passed
- `git diff --check`: passed
- acoustic and variance config inheritance: passed
- previous CodeRabbit standalone-evaluation finding: fixed and confirmed addressed

Real GPU training was not run because no dataset/checkpoint was supplied.

### CODE REVIEW

File: basics/base_task.py [hunk 1]
```diff
@@ -73,6 +75,10 @@ def __init__(self, *args, **kwargs):
         self.phoneme_dictionary = load_phoneme_dictionary()
         self.build_model()
 
+        self.use_ema = hparams.get('ema_enabled', False)
+        self.ema = self.build_ema() if self.use_ema else None
+        self._ema_needs_register = False
```
[the remark this answers is not in the document]
Comment (KakaruHayate) [reply basics/base_task.py:80]: Fixed in aeaf668. EMA now uses a shared evaluation helper that performs any delayed registration before apply(), and the regression test verifies register -> apply ordering for standalone non-fit evaluation.
